src/pybind/mgr/diskprediction_local/predictor.py

```diff
@@ -39,227 +38,158 @@ class DiskFailurePredictor(object):
 
-        for prev, cur in zip(prev_days, curr_days):
-            diff_disk_days.append({attr:(int(cur[attr]) - int(prev[attr]))
-                                   for attr in attr_list})
+        # drop the now-redundant column
+        df = df.drop('ata_smart_attributes', axis=1)
+        return df
 
-        return attr_list, diff_disk_days
 
-    def __get_best_models(self, attr_list):
-        """
-        Find the best model from model list according to given attribute list.
+    def __preprocess(self, disk_days_df):
+        """Scales and transforms input dataframe to feed it to prediction model
 
-        Args:
-            attr_list: All S.M.A.R.T. attributes used in given disk.
+        Arguments:
+            disk_days_df {pandas.DataFrame} -- df where each row holds drive
+                                                features from one day.
 
         Returns:
-            modelpath: The best model for the given attribute list.
-            model_attrlist: 'Ordered' attribute list of the returned model.
-                            Must be aware that SMART attributes is in order.
-
-        Raises:
+            numpy.ndarray -- (n, d) shaped array of n days worth of data and d
+                                features, scaled
         """
-
-        models = self.model_context.keys()
-
-        scores = []
-        for model_name in models:
-            scores.append(sum(attr in attr_list
-                              for attr in self.model_context[model_name]))
-        max_score = max(scores)
-
-        # Skip if too few matched attributes.
-        if max_score < 3:
-            print("Too few matched attributes")
+        # preprocessing may vary across manufactueres. so get manufacturer
+        manufacturer = DiskFailurePredictor.__get_manufacturer(disk_days_df['model_name'].iloc[0]).lower()
+
+        # keep only the features used for prediction for current manufacturer
+        try:
+            disk_days_df = disk_days_df[self.model_context[manufacturer]]
+        except KeyError as e:
+            # TODO: change to log.error
+            print("Either SMART attributes mismatch for hard drive and prediction model,\
+                 or 'model_name' not available in input data")
+            print(e)
             return None
 
-        best_models = {}
-        best_model_indices = [idx for idx, score in enumerate(scores)
-                              if score > max_score - 2]
-        for model_idx in best_model_indices:
-            model_name = list(models)[model_idx]
-            model_path = os.path.join(self.model_dirpath, model_name)
-            model_attrlist = self.model_context[model_name]
-            best_models[model_path] = model_attrlist
+        # scale raw data
+        preprocessor_path = os.path.join(self.model_dirpath, manufacturer + '_preprocessor.joblib')
+        preprocessor = joblib.load(preprocessor_path)
```

Comment:
> Objects are now stored using joblib instead of pickle, as recommended by scikit-learn docs.

as per joblib doc at https://joblib.readthedocs.io/en/latest/persistence.html

> joblib.dump() and joblib.load() are based on the Python pickle serialization model.

and 

> Compatibility of joblib pickles across python versions is not fully supported. 

so we should check current python against the python version used to train the classifier.

Reply:
also, why do you think RGF is better than the SVM used by the original classifier? could you provide some benchmark results?

Reply:
Sure thing. [This](https://hal.archives-ouvertes.fr/hal-01703140/document) research found that GBDTs outperform SVM, and [this](https://www.kdd.org/kdd2016/papers/files/adf0849-botezatuA.pdf) research found that RGF outperforms GBDT (which outperforms SVM in turn, as shown by the former paper too). Then I ran some simple experiments (see [repo](https://github.com/chauhankaranraj/ceph_drive_failure). specifically, [this](https://github.com/chauhankaranraj/ceph_drive_failure/blob/master/clustering_and_model_exploration.ipynb) jupyter notebook) which confirmed the same.

The notebook linked above was about doing a binary classification (fail/nofail). The multiclass classification results (good/warning/bad) for RGF are in [this](https://github.com/chauhankaranraj/ceph_drive_failure/blob/master/multiclass_clf.ipynb) repo (see the "RGF" subheading under the "Train" heading)